doc/design/ops/rnn.md

```diff
@@ -0,0 +1,126 @@
+# RNN design
+This is the design doc of the recurrent neural network operator, 
+and this operator requires that input instances in each mini-batch must have the same length. 
+
+## RNN Algorithm Implementation
+
+<p aligh="center">
+<img src="./images/rnn.jpg"/>
+</p>
+
+The above diagram shows an RNN being unrolled into a full network.
+
+There are several important concepts:
+
+- stepnet, the network execute in every time step 
+- memory, a variable storing state in the current step, which is denoted as $h_t$.
+- pre-memory, the value of state in the previous time step, it can be denoted as $h_{t-1}$.
+- init-memory, the variable to help initialize state in the first time step.
+
+### step scopes
+
+<p aligh="center">
+<img src="./images/rnn.png"/><br/>
+fig 2 the RNN's data flow
+</p>
+
+Each RNN might run one or more steps, each step runs the same step net.
+
+We use `Scope` to store the contexts of all the step times:
+
+- for each step, create a step Scope
+- create all the temporary output variables in the Scope
+- execute the step-net, and each step will have its temporary outputs
+
+After all steps finished, RNNOp will collect the specific outputs of each step and merge them to a larger tensor.
+
+### memory and pre-memory
+a basic RNN is like:
```

Comment:
An RNN step often has some status which needs to be passed to the next step.  These status are known as the memory, which is often referred by the step-net as variables.  A step often needs to refer to the memory value changed by the previous step.  We call it *ex-memory*.

Let's use a simply RNN as an example to explain memory and ex-memory: